odoo/config.py

```diff
@@ -0,0 +1,1478 @@
+
+
+SCAFFOLD_TMPL_DIR = fullpath(__file__).parent.joinpath('cli/templates')
+
+
+def scaffold_builtins():
+    """ List scaffold templates """
+    return [p.name for p in SCAFFOLD_TMPL_DIR.iterdir()]
+
+
+def scaffold_template(rawopt):
+    """ Ensure it is a valid builtin template or template dir """
+    if rawopt in scaffold_builtins():
+        return str(SCAFFOLD_TMPL_DIR.joinpath(rawopt))
+    return str(_check_dir_access(rawopt, os.R_OK))
+
+
+########################################################################
+#                                                                      #
+#                               OPTIONS                                #
+#                                                                      #
+########################################################################
+
+@dataclasses.dataclass(frozen=True)
+class Option:
+    """
+    Option placeholder with all informations needed to find, retrieve
+    and parse the option from various sources.
+    """
+
+    name: str
+    """
+    The option's value will be accessible under this name via the
+    ``Config`` object. The name is also used in various mappings and
+    in config error reports.
+    """
+
+    parse: Callable[[str], Any]
+    """
+    The parsing function used to convert the raw string value to its
+    Python equivalent. The parsing function is also responsible for
+    option validation and sanitization and should raise a ValueError
+    on error.
+    """
+
+    format: Callable[[str], Any] = str
+    """
+    The formatting function used to convert the python parsed option
+    back to the configuration file.
+    """
+
+    required: bool = False
+    """
+    Whether the user have to provide a suitable value via any source,
+    that the is no hardcoded default value should not be used.
+    """
+
+    default: Any = None
+    """
+    The default hardcoded value, ultimate fallback to use if the option
+    has not been given in any other source. The value must have the
+    right python type already.
+    """
+
+    file: bool = True
+    """
+    Whether the option can be read from the configuration file and
+    should be written back on configuration save.
+    """
+
+    envvar: Optional[str] = None
+    """
+    The environment variable where to find the option.
+    """
+
+    args: List[str] = dataclasses.field(default_factory=list)
+    action: str = "store"
+    metavar: Optional[str] = None
+    const: Any = None
+    nargs: Optional[int] = None
+    help: Optional[str] = None
+    """
+    argparse's argument used to dynamically create the cli parser.
+    """
+
+    def __post_init__(self):
+        """ Register every new option in the optionmap """
+        optionmap.setdefault(self.name, self)
+
+Option('addons_path', args=["--addons-path"], parse=CommaList.parser(addons_path), action='append', default=get_default_addons_path(), metavar="DIRPATH", help="specify additional addons paths")
```

Comment:
Maybe Option definitions can be more declarative (or anything so that their definition looks better)? for most of the arguments you can use a dict:

```python
my_opts = [
    {
        'name': 'addons_path',
        'args': [...]
    },
    { ... },
    { ... },
]

for opt_dict in my_opts:
    Option(**opt_dict)
```